Updated CONTRIBUTOR guidelines

davidvossel opened the pull request.

davidvossel commented:
/cc @MarSik @fedepaol @slintes @vladikr @yanirq @cynepco3hahue @fromanirh 

anything missing here?

MarSik commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+* All workflows __must be declarative__. This means the cluster-admin must be able to declare the exact state they want up front, and have our operator handle converging on that state.
+* __No imperative actions__. This means no requirement for a cluster-admin to account for the ordering or timing at which manifests are posted. If our API can't express the desired state up front, then the API is wrong.
+* Limit the Status.Phase enum to only express the state in broad terms. Use events and conditions on the CR to express detailed error information.
+* __Avoid design decisions that result in mutating objects not created and owned by our controller__. Mutating the spec of an object our controller doesn't own is nearly never an acceptable decision. However, there are occasions where adding annotations or labels to an object not owned by the controller makes sense.
```
This one seems to duplicate line 25, it might be intentional though.

davidvossel replied:
yup, fixed

MarSik commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
+* __Avoid performing an apiserver Get()__ request within the reconcile loop if at all possible. Again use informers. There are rare exceptions.
+* Use __finalizers__ in order to perform any cleanup tasks required before a CR is completely removed from etcd.
+* User __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
```
```suggestion
* Use __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
```

davidvossel replied:
fixed

MarSik commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+
+* Use mock clients to simulate reconcile loops.
+* Execute reconcile loop to validate execution under specific conditions.
+* Validate reconcile execution by introspecting API calls that were made during the execution.
```
I would prefer a link to a cook book example here myself.

davidvossel replied:
good idea. I added an example of a test using mock clients

MarSik commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+
+* All functional tests should be written in golang+gomega and exist in the functtests/ directory
+* Functional test scenarios should represent conditions cluster-admins will encounter.
+* Exercise and validate behavior by posting/mutating/deleting manifests and observing the desired state is reached.
```
Any guidelines on timeouts? How long before failing due to: the desired state was not achieved?

davidvossel replied:
added

simon3z commented:
@davidvossel any plan to have code coverage? When available will we add a guideline that PRs shouldn't regress amount of code coverage?

slintes commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+* Never block reconcile execution. ever. This means __no sleeps and no retry loops__
+* All __logic must be idempotent__. This means discovering what has already occurred by inspecting informer cache, and only mutating to converge on desired state.
+* __Never perform an apiserver List()__ request within the reconcile loop. This action is very apiserver intensive. Instead use informers retrieve items from cache.
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
```
isn't the first and last sentence saying the same thing here?

davidvossel replied:
yeah it is redundant. I just wanted to call out the 'spec' explicitly.

fromanirh replied:
I think mentioning the 'spec' explicitely makes the intent of the design principle a little clearer

slintes commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+
+* All workflows __must be declarative__. This means the cluster-admin must be able to declare the exact state they want up front, and have our operator handle converging on that state.
+* __No imperative actions__. This means no requirement for a cluster-admin to account for the ordering or timing at which manifests are posted. If our API can't express the desired state up front, then the API is wrong.
+* Limit the Status.Phase enum to only express the state in broad terms. Use events and conditions on the CR to express detailed error information.
```
I'm tempted to say "avoid phase at all, only use conditions".
https://github.com/kubernetes/kubernetes/issues/7856

davidvossel replied:
very interesting. I tend to agree after reading through that thread. I'll revise this a bit.

davidvossel replied:
revised, and thanks for the suggestion, i like this one.

davidvossel commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,53 @@
+
+* Never block reconcile execution. ever. This means __no sleeps and no retry loops__
+* All __logic must be idempotent__. This means discovering what has already occurred by inspecting informer cache, and only mutating to converge on desired state.
+* __Never perform an apiserver List()__ request within the reconcile loop. This action is very apiserver intensive. Instead use informers retrieve items from cache.
```
fixed

davidvossel commented:
FYI - I addressed feedback and added a new "Deploying Operator on Test Cluster" section that describes how to build and deploy the operator from source using some basic deploy scripts that were introduced earlier today.

davidvossel commented:
> @davidvossel any plan to have code coverage? When available will we add a guideline that PRs shouldn't regress amount of code coverage?

good point. I created an issue so we don't forget to add code coverage reporting. #18

slintes commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,93 @@
+* Use Status.Conditions and Events to represent and record transient states that a CR exists in.
+* Avoid introducing Status.Phase enums if possible. The pitfall of a Phase enum is that it often gets treated as a FSM, when in reality that isn't always the case. [Source](https://github.com/kubernetes/kubernetes/issues/7856)
+
+Limit the Status.Phase enum to only express the state in broad terms. Use events and conditions on the CR to express detailed error information.
```
I think this is a left over from previous version, this is handled in the 2 bullet points above now.
Otherwise lgtm

fromanirh commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,93 @@
+* Use __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
+* Return early to handle errors which re-enqueues the work queue key to be tried again.
+* Do not create logic to sync a key using a recurring time interval. Instead only wake up processing the key when something has changed.
+* Reconcile execution should be broken into two parts, synchronization and status reporting. Synchronization converges the CR on the desired state. Status reporting records the current state of the CR based on the synchronization execution results and the observed state of the cluster. These are two independent functions. __Do not modify status during the synchronization step__.
```
The way I parse this bullet point is: a reconcile step is first synchronization; once that's done, do status reporting. Keep the steps fully separated and independent. Is my understanding correct? If so, maybe let's make this a bit more explicit, like:
```
These are two independent functions. Keep them separated. __Do not modify status during the synchronization step__.
```

davidvossel commented:
@fromanirh @slintes I think I addressed everything now.

davidvossel commented:
/retest

slintes commented on CONTRIBUTING.md:
```diff
@@ -3,16 +3,91 @@
+* __Avoid mutating a cluster object the loop didn't create__. A exception to this would to be adding annotations to an existing object. Mutating a object's Spec that our controller didn't create should be avoided.
+* __Avoid performing an apiserver Get()__ request within the reconcile loop if at all possible. Again use informers. There are rare exceptions.
+* Use __finalizers__ in order to perform any cleanup tasks required before a CR is completely removed from etcd.
+* Use __owner references__ on any objects created and owned by a CR. Owner references ensure objects are automatically garbage collected after a corresponding CR is deleted.
```
following the recent issues in CNV: maybe add that namespaced resource can have namespaced or clusterwide resources as owner, but clusterwide resources are only allowed to have clusterwide owners? Too easy to forget, and doing it wrong results in "magically" at random time deleted resources.

cynepco3hahue commented:
/retest
